api/src/main/java/io/opencensus/stats/ViewData.java

```diff
@@ -53,52 +67,132 @@
-
+    checkWindow(view.getWindow(), windowData);
     Map<List<TagValue>, AggregationData> deepCopy = Maps.newHashMap();
     for (Entry<? extends List<TagValue>, ? extends AggregationData> entry : map.entrySet()) {
+      checkAggregation(view.getAggregation(), entry.getValue(), view.getMeasure());
       deepCopy.put(
           Collections.unmodifiableList(new ArrayList<TagValue>(entry.getKey())), entry.getValue());
     }
-
     return new AutoValue_ViewData(view, Collections.unmodifiableMap(deepCopy), windowData);
   }
 
+  private static void checkWindow(
+      AggregationWindow window, final AggregationWindowData windowData) {
+    window.match(
+        new Function<View.AggregationWindow.Cumulative, Void>() {
+          @Override
+          public Void apply(View.AggregationWindow.Cumulative arg) {
+            checkArgument(
+                windowData instanceof AggregationWindowData.CumulativeData,
+                createErrorMessageForWindow(arg, windowData));
+            return null;
+          }
+        },
+        new Function<View.AggregationWindow.Interval, Void>() {
+          @Override
+          public Void apply(View.AggregationWindow.Interval arg) {
+            checkArgument(
+                windowData instanceof AggregationWindowData.IntervalData,
+                createErrorMessageForWindow(arg, windowData));
+            return null;
+          }
+        },
+        Functions.<Void>throwIllegalArgumentException());
+  }
+
+  private static String createErrorMessageForWindow(
+      AggregationWindow window, AggregationWindowData windowData) {
+    return "AggregationWindow and AggregationWindowData types mismatch. "
+        + "AggregationWindow: "
+        + window
+        + " AggregationWindowData: "
+        + windowData;
+  }
+
+  private static void checkAggregation(
+      final Aggregation aggregation, final AggregationData aggregationData, final Measure measure) {
+    aggregation.match(
+        new Function<Sum, Void>() {
+          @Override
+          public Void apply(Sum arg) {
+            measure.match(
+                new Function<MeasureDouble, Void>() {
+                  @Override
+                  public Void apply(MeasureDouble arg) {
+                    checkArgument(
+                        aggregationData instanceof SumDataDouble,
+                        createErrorMessageForAggregation(aggregation, aggregationData));
+                    return null;
+                  }
+                },
+                new Function<MeasureLong, Void>() {
+                  @Override
+                  public Void apply(MeasureLong arg) {
+                    checkArgument(
+                        aggregationData instanceof SumDataLong,
+                        createErrorMessageForAggregation(aggregation, aggregationData));
+                    return null;
+                  }
+                },
+                Functions.<Void>throwIllegalArgumentException());
```

Comment:
You could throw an AssertionError, because this code is in the API, so it never needs to handle unknown subclasses.

Reply:
Make sense, done.